### PR TITLE
[x86] Add tan intrinsic part 4 

### LINKED ISSUE #70082: [HLSL] implement `tan` intrinsic
Implement HLSL `tan` intrinsic:

https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-tan

### PR DESCRIPTION
**Pull request by farzonl**
This change is an implementation of #87367's investigation on supporting IEEE math operations as intrinsics.
Which was discussed in this RFC:
https://discourse.llvm.org/t/rfc-all-the-math-intrinsics/78294

Much of this change was following how G_FSIN and G_FCOS were used.

Changes:
-  `llvm/docs/GlobalISel/GenericOpcode.rst` - Document the `G_FTAN` opcode 
-  `llvm/docs/LangRef.rst` - Document the tan intrinsic
-  `llvm/include/llvm/Analysis/VecFuncs.def` - Associate the tan intrinsic as a vector function similar to the tanf libcall.
-  `llvm/include/llvm/CodeGen/BasicTTIImpl.h` - Map the tan intrinsic to `ISD::FTAN`
-  `llvm/include/llvm/CodeGen/ISDOpcodes.h` - Define ISD opcodes for `FTAN` and `STRICT_FTAN`
-  `llvm/include/llvm/IR/Intrinsics.td` - Create the tan intrinsic
-  `llvm/include/llvm/IR/RuntimeLibcalls.def` - Define tan libcall mappings
-  `llvm/include/llvm/Target/GenericOpcodes.td` - Define the `G_FTAN` Opcode
-  `llvm/include/llvm/Support/TargetOpcodes.def` - Create a `G_FTAN` Opcode handler
-  `llvm/include/llvm/Target/GlobalISel/SelectionDAGCompat.td`  - Map `G_FTAN` to `ftan`
-  `llvm/include/llvm/Target/TargetSelectionDAG.td` - Define `ftan`, `strict_ftan`,  and `any_ftan` and map them to the ISD opcodes for `FTAN` and `STRICT_FTAN`
-  `llvm/lib/Analysis/VectorUtils.cpp` - Associate the tan intrinsic as a vector intrinsic
-  `llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp` Map the tan intrinsic to  `G_FTAN` Opcode
-  `llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp` - Add `G_FTAN` to the list of  floating point math operations also associate `G_FTAN` with the `TAN_F` runtime lib.
- `llvm/lib/CodeGen/GlobalISel/Utils.cpp` - More floating point math operation common behaviors.
-  llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp - List the function expansion operations for `FTAN` and `STRICT_FTAN`. Also define both opcodes in `PromoteNode`.
- `llvm/lib/CodeGen/SelectionDAG/LegalizeFloatTypes.cpp` - More `FTAN` and `STRICT_FTAN` handling in the legalizer
-  `llvm/lib/CodeGen/SelectionDAG/LegalizeTypes.h` - Define `SoftenFloatRes_FTAN` and `ExpandFloatRes_FTAN`.
- `llvm/lib/CodeGen/SelectionDAG/LegalizeVectorOps.cpp` - Define `FTAN` as a legal vector operation.
-  `llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp` - Define `FTAN` as a legal vector operation.
-  `llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp` - define tan as an intrinsic that doesn't return NaN.
-  `llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp` Map `LibFunc_tan`, `LibFunc_tanf`, and `LibFunc_tanl` to `ISD::FTAN`. Map `Intrinsic::tan` to `ISD::FTAN` and add selection dag handling for `Intrinsic::tan`.
-  `llvm/lib/CodeGen/SelectionDAG/SelectionDAGDumper.cpp` - Define  `ftan` and `strict_ftan` names for the equivalent ISD  opcodes.
- `llvm/lib/CodeGen/TargetLoweringBase.cpp` -Define a Tan128 libcall and ISD::FTAN as a target lowering action.
-  `llvm/lib/Target/X86/X86ISelLowering.cpp` - Add x86_64 lowering for tan intrinsic

resolves https://github.com/llvm/llvm-project/issues/70082

### CODE REVIEW

**Comment by tschuett** [comment]
Nit: the `declare` is not necessary anymore. You touched the IRTranslator without a test at, e.g., https://github.com/llvm/llvm-project/blob/main/llvm/test/CodeGen/AArch64/GlobalISel/arm64-irtranslator.ll

**Comment by farzonl** [comment]
> Nit: the `declare` is not necessary anymore. You touched the IRTranslator without a test at, e.g., https://github.com/llvm/llvm-project/blob/main/llvm/test/CodeGen/AArch64/GlobalISel/arm64-irtranslator.ll

The aarch64 backend tag is a misnomer because adding `G_FTAN` caused the legalizer to complain in this arm64 test case: llvm/test/CodeGen/AArch64/GlobalISel/legalizer-info-validation.mir

I could add
```
declare float @llvm.tan.f32(float)
define float @test_tan_f32(float %x) {
  ; CHECK-LABEL: name:            test_tan_f32
  ; CHECK: %{{[0-9]+}}:_(s32) = G_FTAN %{{[0-9]+}}
  %y = call float @llvm.tan.f32(float %x)
  ret float %y
}
```
But I feel like this should be added when a tan intrinsic lowering is done for arm64 in llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

**Comment by tschuett** [comment]
Firstly, it is a nit. The IRTranslator translates LLVM-IR into GMIR for GlobalIsel. Testing the IRTranslator is independent of any SDAG changes.

File: llvm/test/CodeGen/AArch64/GlobalISel/legalizer-info-validation.mir
Comment (tschuett) [inline llvm/test/CodeGen/AArch64/GlobalISel/legalizer-info-validation.mir:677]: It means that the AArch64 GlobalIsel legalizer does not say anything about G_FTAN.

Comment (farzonl) [reply llvm/test/CodeGen/AArch64/GlobalISel/legalizer-info-validation.mir:677]: I know. There is no way to do a `G_FTAN` that isn't exposed across all backends.

File: llvm/lib/Target/X86/X86ISelLowering.cpp
Comment (topperc) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:658]: I want to tell you to match the existing code that uses extra spaces to keep things in columns, but I guess clang-format will complain.

File: llvm/test/CodeGen/AArch64/GlobalISel/legalizer-info-validation.mir
Comment (farzonl) [reply llvm/test/CodeGen/AArch64/GlobalISel/legalizer-info-validation.mir:677]: I am tracking changes to support arm64 here: https://github.com/farzonl/llvm-project/commit/aade6dde06f686e35f5c526cd915ba3b76380d16

It will have to come after this PR merges.

File: llvm/lib/Target/X86/X86ISelLowering.cpp
Comment (farzonl) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:658]: Exactly, the format is changed because of clang-format. Without changing the github action, I don't see how I can preserve the columns spacing.

**Comment by farzonl** [comment]
@RKSimon @topperc When you have time could I get a review. Thank you!

**Comment by farzonl** [comment]
@RKSimon @topperc Sorry to ping again, if you have time this week, I would really appreciate a review! Thanks in advance.

File: llvm/include/llvm/Analysis/VecFuncs.def
Comment (topperc) [inline llvm/include/llvm/Analysis/VecFuncs.def:57]: What about the other places in this file that use "tanf"?

File: llvm/lib/Target/X86/X86ISelLowering.cpp
Comment (topperc) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:741]: I think this should be unneeded due to TargetLoweringBase.cpp defaulting to Expand.

File: llvm/lib/Target/X86/X86ISelLowering.cpp
Comment (topperc) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:734]: I think this should be unneeded due to TargetLoweringBase.cpp defaulting to Expand.

File: llvm/test/CodeGen/X86/llvm.tan.ll
Comment (RKSimon) [inline llvm/test/CodeGen/X86/llvm.tan.ll:34]: add f80 test?

File: llvm/test/CodeGen/X86/llvm.tan.ll
Comment (RKSimon) [inline llvm/test/CodeGen/X86/llvm.tan.ll:4]: add nounwind to get rid of the cfi noise

File: llvm/lib/Target/X86/X86ISelLowering.cpp
Comment (RKSimon) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:941]: indentation

File: llvm/lib/CodeGen/SelectionDAG/LegalizeTypes.h
Comment (RKSimon) [inline llvm/lib/CodeGen/SelectionDAG/LegalizeTypes.h:652]: use the same indentation as the other declarations

File: llvm/lib/CodeGen/SelectionDAG/LegalizeFloatTypes.cpp
Comment (RKSimon) [inline llvm/lib/CodeGen/SelectionDAG/LegalizeFloatTypes.cpp:1413]: use the same indentation as the rest of the cases

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp
Comment (RKSimon) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:6712]: indentation

Comment (farzonl) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:6712]: I can't, clan format github action will fail.

Comment (arsenm) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:6712]: I think we need to fix this once and for all by turning all of these horrible case, assign, break patterns into separate helper functions that just return the value. This happens on every DAG opcode patch

File: llvm/include/llvm/Analysis/VecFuncs.def
Comment (farzonl) [reply llvm/include/llvm/Analysis/VecFuncs.def:57]: I think I have a robust answer for you: 

For now we are going to skip 
`DARWIN_LIBSYSTEM_M_VECFUNCS` as those are aarch64 specific
```cpp
TLI_DEFINE_VECFUNC("tan", "_simd_tan_d2", FIXED(2), "_ZGV_LLVM_N2v")
TLI_DEFINE_VECFUNC("llvm.tan.f64", "_simd_tan_d2", FIXED(2), "_ZGV_LLVM_N2v")
TLI_DEFINE_VECFUNC("tanf", "_simd_tan_f4", FIXED(4), "_ZGV_LLVM_N4v")
TLI_DEFINE_VECFUNC("llvm.tan.f32", "_simd_tan_f4", FIXED(4), "_ZGV_LLVM_N4v")
```

We are going to skip `MASSV_VECFUNCS` because these are powerpc specific so no
```cpp
TLI_DEFINE_VECFUNC("llvm.tan.f64", "__tand2", FIXED(2), "_ZGV_LLVM_N2v")
TLI_DEFINE_VECFUNC("llvm.tan.f32", "__tanf4", FIXED(4), "_ZGV_LLVM_N4v")
```

We are going to skip `SLEEFGNUABI_VF2_VECFUNCS` so no `ZGVnN2v_tan` or `_ZGVnN4v_tanf`, these appear to be aarch64\NEON specific 

We are going to skip `SLEEFGNUABI_SCALABLE_VECFUNCS` so no `_ZGVsMxv_tan` or `_ZGVsMxv_tanf` these appear to be aarch64\SVE specific 
```cpp
TLI_DEFINE_VECFUNC("llvm.tan.f64", "_ZGVsMxv_tan", SCALABLE(2), MASKED, "_ZGVsMxv")
TLI_DEFINE_VECFUNC("llvm.tan.f32", "_ZGVsMxv_tanf", SCALABLE(4), MASKED, "_ZGVsMxv")
```

We are going to skip `ARMPL_VECFUNCS` as these are another type of aarch64 neon instruction so no: 

```cpp
TLI_DEFINE_VECFUNC("llvm.tan.f64", "armpl_vtanq_f64", FIXED(2), NOMASK, "_ZGV_LLVM_N2v")
TLI_DEFINE_VECFUNC("llvm.tan.f32", "armpl_vtanq_f32", FIXED(4), NOMASK, "_ZGV_LLVM_N4v")
TLI_DEFINE_VECFUNC("llvm.tan.f64", "armpl_svtan_f64_x", SCALABLE(2), MASKED, "_ZGVsMxv")
TLI_DEFINE_VECFUNC("llvm.tan.f32", "armpl_svtan_f32_x", SCALABLE(4), MASKED, "_ZGVsMxv")
```

We are going to add AMDLIBM_VECFUNCS LIBMVEC_X86_VECFUNCS and _SVML_VECFUNCS the tests indicate these are all x86

so expect to see:

```cpp
TLI_DEFINE_VECFUNC("llvm.tan.f32", "amd_vrs16_tanf", FIXED(16), NOMASK, "_ZGV_LLVM_N16v")
TLI_DEFINE_VECFUNC("llvm.tan.f32", "amd_vrs8_tanf", FIXED(8), NOMASK, "_ZGV_LLVM_N8v")
TLI_DEFINE_VECFUNC("llvm.tan.f32", "amd_vrs4_tanf", FIXED(4), NOMASK, "_ZGV_LLVM_N4v")
TLI_DEFINE_VECFUNC("llvm.tan.f64", "amd_vrd8_tan", FIXED(8), NOMASK, "_ZGV_LLVM_N8v")
TLI_DEFINE_VECFUNC("llvm.tan.f64", "amd_vrd4_tan", FIXED(4), NOMASK, "_ZGV_LLVM_N4v")
TLI_DEFINE_VECFUNC("llvm.tan.f64", "amd_vrd2_tan", FIXED(2), NOMASK, "_ZGV_LLVM_N2v")
```

and:
```cpp
TLI_DEFINE_VECFUNC("tan", "__svml_tan2", FIXED(2), "_ZGV_LLVM_N2v")
TLI_DEFINE_VECFUNC("tan", "__svml_tan4", FIXED(4), "_ZGV_LLVM_N4v")
TLI_DEFINE_VECFUNC("tan", "__svml_tan8", FIXED(8), "_ZGV_LLVM_N8v")

TLI_DEFINE_VECFUNC("tanf", "__svml_tanf4", FIXED(4), "_ZGV_LLVM_N4v")
TLI_DEFINE_VECFUNC("tanf", "__svml_tanf8", FIXED(8), "_ZGV_LLVM_N8v")
TLI_DEFINE_VECFUNC("tanf", "__svml_tanf16", FIXED(16), "_ZGV_LLVM_N16v")

TLI_DEFINE_VECFUNC("llvm.tan.f64", "__svml_tan2", FIXED(2), "_ZGV_LLVM_N2v")
TLI_DEFINE_VECFUNC("llvm.tan.f64", "__svml_tan4", FIXED(4), "_ZGV_LLVM_N4v")
TLI_DEFINE_VECFUNC("llvm.tan.f64", "__svml_tan8", FIXED(8), "_ZGV_LLVM_N8v")

TLI_DEFINE_VECFUNC("llvm.tan.f32", "__svml_tanf4", FIXED(4), "_ZGV_LLVM_N4v")
TLI_DEFINE_VECFUNC("llvm.tan.f32", "__svml_tanf8", FIXED(8), "_ZGV_LLVM_N8v")
TLI_DEFINE_VECFUNC("llvm.tan.f32", "__svml_tanf16", FIXED(16), "_ZGV_LLVM_N16v")
```

and:
```cpp
TLI_DEFINE_VECFUNC("tan", "_ZGVbN2v_tan", FIXED(2), "_ZGV_LLVM_N2v")
TLI_DEFINE_VECFUNC("tan", "_ZGVdN4v_tan", FIXED(4), "_ZGV_LLVM_N4v")

TLI_DEFINE_VECFUNC("tanf", "_ZGVbN4v_tanf", FIXED(4), "_ZGV_LLVM_N4v")
TLI_DEFINE_VECFUNC("tanf", "_ZGVdN8v_tanf", FIXED(8), "_ZGV_LLVM_N8v")

TLI_DEFINE_VECFUNC("llvm.tan.f64", "_ZGVbN2v_tan", FIXED(2), "_ZGV_LLVM_N2v")
TLI_DEFINE_VECFUNC("llvm.tan.f64", "_ZGVdN4v_tan", FIXED(4), "_ZGV_LLVM_N4v")

TLI_DEFINE_VECFUNC("llvm.tan.f32", "_ZGVbN4v_tanf", FIXED(4), "_ZGV_LLVM_N4v")
TLI_DEFINE_VECFUNC("llvm.tan.f32", "_ZGVdN8v_tanf", FIXED(8), "_ZGV_LLVM_N8v")
```

and tests for these three

File: llvm/lib/CodeGen/SelectionDAG/LegalizeTypes.h
Comment (farzonl) [reply llvm/lib/CodeGen/SelectionDAG/LegalizeTypes.h:652]: I can add some clang-format off\ons in the code base to handle this.

File: llvm/lib/CodeGen/SelectionDAG/LegalizeFloatTypes.cpp
Comment (farzonl) [reply llvm/lib/CodeGen/SelectionDAG/LegalizeFloatTypes.cpp:1413]: will add clang-format off\ons

File: llvm/test/CodeGen/X86/llvm.tan.ll
Comment (farzonl) [reply llvm/test/CodeGen/X86/llvm.tan.ll:32]: done

File: llvm/lib/Target/X86/X86ISelLowering.cpp
Comment (farzonl) [reply llvm/lib/Target/X86/X86ISelLowering.cpp]: A few questions. Firs if expand is default then why do we have so many Expands in this file? For example, in this block there is a sin\cos expand. 

Outside of the two cases you called out here there are other expands added in this file for tan.  Are you saying we should only be calling `setOperationAction` if we are doing `Action` or `LibCall`?

**Comment by farzonl** [comment]
@RKSimon all issues you raised have been addressed can I get another review?
@topperc There is one outstanding issue I want clarification on via `setOperationAction` usage in Expand cases, all other issues are addressed.

File: llvm/lib/Target/X86/X86ISelLowering.cpp
Comment (topperc) [reply llvm/lib/Target/X86/X86ISelLowering.cpp]: sin/cos default to `Legal` because there are no calls to `setOperationAction` for them in TargetLoweringBase.cpp. The underlying array that `setOperationAction` writes to defaults to `Legal` in the constructor.

This patch modifies code in TargetLoweringBase.cpp to Expand ISD::TAN for f32/f64/f128. That makes any setOperationAction to `Expand` for ISD::TAN for f32/f64/f128 in X86ISelLowering.cpp redundant. Other types like f16, f80 or vectors still need to be done explicitly in X86ISelLowering.cpp.

```
  setOperationAction({ISD::FCBRT, ISD::FLOG, ISD::FLOG2, ISD::FLOG10, ISD::FEXP,
                      ISD::FEXP2, ISD::FEXP10, ISD::FFLOOR, ISD::FNEARBYINT,
                      ISD::FCEIL, ISD::FRINT, ISD::FTRUNC, ISD::LROUND,
                      ISD::LLROUND, ISD::LRINT, ISD::LLRINT, ISD::FROUNDEVEN,
                      ISD::FTAN},
                     {MVT::f32, MVT::f64, MVT::f128}, Expand);
```

File: llvm/lib/Target/X86/X86ISelLowering.cpp
Comment (farzonl) [reply llvm/lib/Target/X86/X86ISelLowering.cpp]: done

File: llvm/lib/Target/X86/X86ISelLowering.cpp
Comment (farzonl) [reply llvm/lib/Target/X86/X86ISelLowering.cpp]: Thank you for the detailed answer. I noticed two other f32\f64 tan cases so removed those as well.

File: llvm/test/CodeGen/X86/llvm.tan.ll
Comment (RKSimon) [inline llvm/test/CodeGen/X86/llvm.tan.ll:33]: This should be x86_fp80 not double

Comment (farzonl) [reply llvm/test/CodeGen/X86/llvm.tan.ll]: fixed

**Comment by farzonl** [comment]
@RKSimon @topperc there are no more outstanding issues on this PR. Please let me know if this is ready for final review.

**Review by RKSimon — APPROVED** [review]
LGTM - cheers

**Review by llvm-beanz — APPROVED** [review]

**Pull request merged**
